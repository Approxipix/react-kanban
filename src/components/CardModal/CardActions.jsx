import React, { Component } from 'react';
import CardChecklistAdd from '../Checklist/CheckListAdd';
import CardDelete from '../Card/CardDelete';
import LabelAdd from '../Label/LabelAdd';
import CardTooltip from './CardTooltip';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';
import PropTypes from "prop-types";

const Wrapper = styled.div`
  position: relative;
  padding: .5rem 0;
  :not(:last-child) {
    margin-bottom: 1rem;
  }
`;

const Title = styled.h4`
  margin-bottom: .5rem;
  font-size: .75rem;
  color: #6b778c;
  font-weight: 500;
  text-transform: uppercase;
`;

const ActionList = styled.ul`
  display: flex;
  flex-direction: column;
`;
ActionList.displayName = 'ActionList';

const ActionItem = styled.li`
  position: relative;
  &:not(:last-child) {
    margin-bottom: .5rem;
  }
`;

const Link = styled.a`
  display: flex;
  align-items: center;
  width: 100%;
  padding: .5rem .75rem;
  font-size: .875rem;
  font-weight: 600;
  border-radius: .2rem;
  text-align: left;
  background-color: #ebecf0;
  box-shadow: 0 1px 0 0 rgba(9, 30, 66, .13);
  transition: all .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: #dfe1e6;
    box-shadow: 0 1px 0 0 rgba(9, 30, 66, .25);
  }
`;
Link.displayName = 'Link';

const Icon = styled.div`
  margin-right: .5rem;
`;

class CardActions extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpened: null
    }
  }

  toggleIsOpened = (key) => {
    this.setState({
      isOpened: key !== this.state.isOpened ? key : null
    })
  };

  render() {
    const { listId, card, title, actions } = this.props;
    const { isOpened } = this.state;
    let actionItems = [
      {
        title: 'Label',
        icon: 'tag',
        component: (
          <LabelAdd
            cardId={card._cardId}
            cardLabels={card.cardLabels}
          />
        )
      },
      {
        title: 'CheckList',
        icon: 'check-square',
        component: (
          <CardChecklistAdd
            cardId={card._cardId}
            toggleTooltip={() => this.toggleIsOpened('CheckList')}
          />
        )
      },
      {
        title: 'Delete',
        icon: 'trash',
        component: (
          <CardDelete
            cardId={card._cardId}
            listId={listId}
            toggleTooltip={() => this.toggleIsOpened('Delete')}
          />
        )
      },
    ];
    actionItems = actionItems.filter(item => actions.some(title => title === item.title));

    return (
      <Wrapper>
        <Title>
          {title}
        </Title>
        <ActionList>
          {actionItems.map((action, index) => (
            <ActionItem key={index}>
              <Link onClick={() => this.toggleIsOpened(action.title)}>
                <Icon>
                  <FontAwesomeIcon icon={action.icon} />
                </Icon>
                {action.title}
              </Link>
              <CardTooltip
                title={action.title}
                isOpened={isOpened === action.title}
                toggleTooltip={() => this.toggleIsOpened(action.title)}
                body={action.component}
              />
            </ActionItem>
          ))}
        </ActionList>
      </Wrapper>
    )
  }
}

CardActions.defaultProps = {
  card: {},
  listId: '',
  title: '',
  actions: [],
};

CardActions.propTypes = {
  card: PropTypes.shape({
    _cardId: PropTypes.string.isRequired,
    title: PropTypes.string,
    description: PropTypes.string,
    checkLists: PropTypes.arrayOf(PropTypes.string),
    cardLabels: PropTypes.array,
  }).isRequired,
  listId: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  actions: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default CardActions;
