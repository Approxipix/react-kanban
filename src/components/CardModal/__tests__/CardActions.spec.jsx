import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardActions from '../CardActions';

describe('<CardActions>', () => {
  let component;

  beforeEach(() => {
    const props = {
      card: {
        _cardId: "CardID",
        cardLabels: [],
      },
      listId: "ListID",
      title: "Test Title",
      actions: ["Label", "CheckList"],
    };
    const state = {
      isOpened: 'Label'
    };

    component = shallow(<CardActions {...props} />);
    component.setState({...state})
  });

  it('should renders without crashing given the required props', () => {
    expect(toJson(component)).toMatchSnapshot()
  });

  it('should render a component for each action', () => {
    expect(component.find('ActionList').children().length).toEqual(2)
  });

  it('checks if toggleIsOpened sets isOpened correctly', () => {
    expect(component.instance().state.isOpened).toEqual('Label');

    component.instance().toggleIsOpened('CheckList');
    expect(component.instance().state.isOpened).toEqual('CheckList');

    component.instance().toggleIsOpened('CheckList');
    expect(component.instance().state.isOpened).toEqual(null);
  });

  it("should call toggleIsOpened after click on link", () => {
    const spy = jest.spyOn(component.instance(), "toggleIsOpened");

    component.find('Link').last().simulate('click', 'CheckList');

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith('CheckList');
  });
});
