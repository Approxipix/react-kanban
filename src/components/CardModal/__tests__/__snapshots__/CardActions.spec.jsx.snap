// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardActions> should renders without crashing given the required props 1`] = `
<styled.div>
  <styled.h4>
    Test Title
  </styled.h4>
  <ActionList>
    <styled.li
      key="0"
    >
      <Link
        onClick={[Function]}
      >
        <styled.div>
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="tag"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.div>
        Label
      </Link>
      <CardTooltip
        body={
          <Memo(Connect(LabelAdd))
            cardId="CardID"
            cardLabels={Array []}
          />
        }
        isOpened={true}
        title="Label"
        toggleTooltip={[Function]}
      />
    </styled.li>
    <styled.li
      key="1"
    >
      <Link
        onClick={[Function]}
      >
        <styled.div>
          <FontAwesomeIcon
            border={false}
            className=""
            fixedWidth={false}
            flip={null}
            icon="check-square"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            swapOpacity={false}
            symbol={false}
            title=""
            transform={null}
          />
        </styled.div>
        CheckList
      </Link>
      <CardTooltip
        body={
          <Memo(Connect(CheckListAdd))
            cardId="CardID"
            toggleTooltip={[Function]}
          />
        }
        isOpened={false}
        title="CheckList"
        toggleTooltip={[Function]}
      />
    </styled.li>
  </ActionList>
</styled.div>
`;
